osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/DlmsPushNotificationDecoderTest.java

```diff
@@ -0,0 +1,388 @@
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+import static org.opensmartgridplatform.dto.valueobjects.smartmetering.AlarmTypeDto.COMMUNICATION_ERROR_M_BUS_CHANNEL_4;
+import static org.opensmartgridplatform.dto.valueobjects.smartmetering.AlarmTypeDto.FRAUD_ATTEMPT_M_BUS_CHANNEL_1;
+import static org.opensmartgridplatform.dto.valueobjects.smartmetering.AlarmTypeDto.PHASE_OUTAGE_DETECTED_L1;
+import static org.opensmartgridplatform.dto.valueobjects.smartmetering.AlarmTypeDto.REPLACE_BATTERY;
+
+import java.util.Set;
+
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.channel.Channel;
+import org.jboss.netty.channel.ChannelHandlerContext;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.opensmartgridplatform.adapter.protocol.dlms.infra.networking.DlmsPushNotificationDecoder.DecodingState;
+import org.opensmartgridplatform.dlms.DlmsPushNotification;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.AlarmTypeDto;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DlmsPushNotificationDecoderTest {
+
+    private static final int EQUIPMENT_IDENTIFIER_LENGTH = 17;
+    private static final int LOGICAL_NAME_LENGTH = 6;
+    private static final int COMMA_LENGTH = 1;
+
+    private static final int SMR5_NUMBER_OF_BYTES_FOR_ADDRESSING = 8;
+    private static final int SMR5_NUMBER_OF_BYTES_FOR_INVOKE_ID = 4;
+    private static final int SMR5_NUMBER_OF_BYTES_FOR_DATETIME = 7;
+
+    private static final String PUSH_SCHEDULER_TRIGGER = "Push scheduler";
+    private static final String PUSH_CSD_TRIGGER = "Push csd wakeup";
+    private static final String PUSH_SMS_TRIGGER = "Push sms wakeup";
+    private static final String PUSH_ALARM_TRIGGER = "Push alarm monitor";
+
+    private static final byte[] SCHEDULER_OBISCODE_BYTES = new byte[] { 0x00, 0x00, 0x0F, 0x00, 0x04, (byte)0xFF };
+    private static final byte[] SCHEDULER_SETUP_OBISCODE_BYTES = new byte[] { 0x00, 0x00, 0x19, 0x09, 0x00, (byte)0xFF };
+    private static final byte[] CSD_OBISCODE_BYTES = new byte[] { 0x00, 0x00, 0x02, 0x02, 0x00, (byte) 0xFF };
+    private static final byte[] SMS_OBISCODE_BYTES = new byte[] { 0x00, 0x00, 0x02, 0x03, 0x00, (byte) 0xFF };
+    private static final byte[] ALARM_OBISCODE_BYTES = new byte[] { 0x00, 0x01, 0x19, 0x09, 0x00, (byte) 0xFF };
+    private static final byte[] SMR5_EXTERNAL_TRIGGER_OBISCODE_BYTES = new byte[] { 0x00, 0x00, 0x02, 0x02, 0x00,
+            (byte) 0xFF };
+
+    private static final String IDENTIFIER = "EXXXX123456789012";
+    private static final byte COMMA = 0x2C;
+
+    private static final int STRUCTURE = 0x02;
+    private static final int OCTET_STRING = 0x09;
+    private static final int DOUBLE_LONG_UNSIGNED = 0x06;
+
+    private DlmsPushNotificationDecoder decoder;
+
+    @Mock
+    ChannelHandlerContext ctx;
+
+    @Mock
+    Channel channel;
```

Comment:
Can these package-private fields become private?

Reply:
Of course :)